src/base/TypeUtil.ml

```diff
@@ -65,23 +65,31 @@ module type MakeTEnvFunctor = functor (Q : QualifiedTypes) (R : Rep) -> sig
   module TEnv : sig
     type t
 
+    type restore
+
     (* Make new type environment *)
-    val mk : t
+    val mk : unit -> t
 
     (* Add to type environment *)
-    val addT : t -> R.rep TUIdentifier.t -> TUType.t -> t
+    val addT : t -> R.rep TUIdentifier.t -> TUType.t -> restore list
 
     (* Add to many type bindings *)
-    val addTs : t -> (R.rep TUIdentifier.t * TUType.t) list -> t
+    val addTs : t -> (R.rep TUIdentifier.t * TUType.t) list -> restore list
```

Comment:
You could also have the return type be `restore` rather than `restore list` here. The `restore` object returned by the function corresponds to the reversal of all the additions, but as a single `restore` operation.

Reply:
Similar observations about all the other `TEnv` functions.